File: framework/languages/java/samples/java/TicTacToe/src/main/java/systems/zlink/samples/tictactoe/server/play/sessions/PlaySession.java
package systems.zlink.samples.tictactoe.server.play.sessions;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.framework.streams.ZLinkStreamHeader;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.play.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.gamespots.TicTacToeGameDirectory;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateReq;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateRes;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameReq;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkReq;
import systems.zlink.stream.connector.ZLinkStreamEncodedPayload;
import systems.zlink.stream.connector.json.ZLinkStreamJson;

public final class PlaySession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkActorManager actors;
    private String actorId;
    private PlayActor playActor;
    private ZLinkSessionActor actor;

    public PlaySession(
        ZLinkSessionContext context,
        ZLinkActorManager actors) {
        this.context = context;
        this.actors = actors;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnectedAsync() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnectedAsync() {
        actorId = null;
        playActor = null;
        actor = null;
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onErrorAsync(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatchAsync(ZLinkStreamHeader header, Message payload) {
        try {
            if (header.packetName().equals("AuthenticateReq")) {
                AuthenticateReq request = decode(header, payload, AuthenticateReq.class);
                actorId = request.actorId();
                return actors.getOrCreateAsync(actorId, SampleNames.PlayActor)
                    .thenCompose(created -> {
                        playActor = (PlayActor) created;
                        return context.actors().bindAsync(created);
                    })
                    .thenCompose(bound -> {
                        actor = bound;
                        return context.client()
                            .reply(new AuthenticateRes(actorId))
                            .submitAsync();
                    });
            }
            if (header.packetName().equals("JoinGameReq")) {
                JoinGameReq request = decode(header, payload, JoinGameReq.class);
                PlayActor playActor = requirePlayActor();
                playActor.joinGame(request.gameId());
                return context.client()
                    .reply(TicTacToeGameDirectory.get(request.gameId())
                        .join(playActor.actorId()))
                    .submitAsync();
            }
            if (header.packetName().equals("PlaceMarkReq")) {
                PlaceMarkReq request = decode(header, payload, PlaceMarkReq.class);
                PlayActor playActor = requirePlayActor();
                return context.client()
                    .reply(TicTacToeGameDirectory.get(playActor.gameId())
                        .placeMark(playActor.actorId(), request.cell()))
                    .submitAsync();
            }
            return requireActor().relayAsync(header, payload);
        } catch (RuntimeException ex) {
            return CompletableFuture.failedFuture(ex);
        }
    }

    private ZLinkSessionActor requireActor() {
        if (actorId == null || actorId.isBlank() || actor == null) {
            throw new IllegalStateException("AuthenticateReq is required before play packets.");
        }
        return actor;
    }

    private PlayActor requirePlayActor() {
        requireActor();
        if (playActor == null) {
            throw new IllegalStateException("bound PlayActor is missing.");
        }
        return playActor;
    }

    private static <T> T decode(ZLinkStreamHeader header, Message payload, Class<T> type) {
        return ZLinkStreamJson.decode(
            new ZLinkStreamEncodedPayload(header.packetName(), payload, header.metadata()),
            type);
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/sessions/PlaySession.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.sessions

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.contracts.messaging.Message
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.streams.ZLinkSession
import systems.zlink.framework.streams.ZLinkSessionActor
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.framework.streams.ZLinkStreamHeader
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.play.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.gamespots.TicTacToeGameDirectory
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlaceMarkReq
import systems.zlink.stream.connector.ZLinkStreamEncodedPayload
import systems.zlink.stream.connector.json.ZLinkStreamJson

class PlaySession(
    private val context: ZLinkSessionContext,
    private val actors: ZLinkActorManager,
) : ZLinkSession {
    private var actorId: String? = null
    private var playActor: PlayActor? = null
    private var actor: ZLinkSessionActor? = null

    override fun context(): ZLinkSessionContext = context
    override fun onConnectedAsync(): CompletionStage<Void> = CompletableFuture.completedFuture(null)
    override fun onDisconnectedAsync(): CompletionStage<Void> {
        actorId = null
        playActor = null
        actor = null
        return CompletableFuture.completedFuture(null)
    }
    override fun onErrorAsync(error: ZLinkStreamError): CompletionStage<Void> =
        CompletableFuture.completedFuture(null)

    override fun onDispatchAsync(header: ZLinkStreamHeader, payload: Message): CompletionStage<Void> =
        try {
            if (header.packetName() == "AuthenticateReq") {
                val request = decode(header, payload, AuthenticateReq::class.java)
                actorId = request.actorId
                actors.getOrCreateAsync(request.actorId, SampleNames.PlayActor)
                    .thenCompose { created ->
                        playActor = created as PlayActor
                        context.actors().bindAsync(created)
                    }
                    .thenCompose { bound ->
                        actor = bound
                        context.client().reply(AuthenticateRes(request.actorId)).submitAsync()
                    }
            } else if (header.packetName() == "JoinGameReq") {
                val request = decode(header, payload, JoinGameReq::class.java)
                val currentActor = requirePlayActor()
                currentActor.joinGame(request.gameId)
                context.client()
                    .reply(TicTacToeGameDirectory.get(request.gameId).join(currentActor.actorId()))
                    .submitAsync()
            } else if (header.packetName() == "PlaceMarkReq") {
                val request = decode(header, payload, PlaceMarkReq::class.java)
                val currentActor = requirePlayActor()
                context.client()
                    .reply(
                        TicTacToeGameDirectory.get(currentActor.gameId)
                            .placeMark(currentActor.actorId(), request.cell),
                    )
                    .submitAsync()
            } else {
                requireActor().relayAsync(header, payload)
            }
        } catch (ex: RuntimeException) {
            CompletableFuture.failedFuture(ex)
        }

    private fun requireActor(): ZLinkSessionActor =
        actor?.takeIf { !actorId.isNullOrBlank() }
            ?: throw IllegalStateException("AuthenticateReq is required before play packets.")

    private fun requirePlayActor(): PlayActor {
        requireActor()
        return playActor ?: throw IllegalStateException("bound PlayActor is missing.")
    }

    private fun <T> decode(header: ZLinkStreamHeader, payload: Message, type: Class<T>): T =
        ZLinkStreamJson.decode(
            ZLinkStreamEncodedPayload(header.packetName(), payload, header.metadata()),
            type,
        )
}
